raft.go

```diff
@@ -652,6 +702,75 @@ func (r *Raft) verifyLeader(v *verifyFuture) {
 	}
 }
 
+// leadershipTransfer is doing the heavy lifting for the leadership transfer.
+func (r *Raft) leadershipTransfer(future *leadershipTransferFuture) {
+
+	// Step 1: set this field which stops this leader from responding to any client requests.
+	r.leaderState.leadershipTransferInProgress = true
+
+	// Step 2: make sure the target server is up to date
+	s, ok := r.leaderState.replState[future.ID]
+	if !ok {
+		err := fmt.Errorf("cannot find replication state for %v, aborting leadership transfer", future)
+		r.logger.Printf("[DEBUG] raft: %v", err)
+		future.respond(err)
+		return
+	}
+
+	for s.nextIndex <= r.getLastIndex() {
+		r.logger.Printf("[DEBUG] raft: waiting for replication to catch up before sending TimeoutNow")
+		err := &deferError{}
+		err.init()
+		s.triggerDeferErrorCh <- err
+		if err.Error() != nil {
+			r.logger.Printf("[DEBUG] raft: %v", err.Error())
+			future.respond(err.Error())
+			return
+		}
+	}
+
+	// Step 3: reset leader lease. Since a new leader might be choosen in a bit,
+	// we don't want this leader to have the right to answer read request based on
+	// this lease.
+	r.leaderState.lease = time.After(0)
```

Comment:
Hmm I'm not sure this is right for a very confusing reason!

The concept of a lease in our leader implementation is not really quite the same as the lease mechanism described in `6.4.1 Using clocks to reduce messaging for read-only queries` of the thesis.

That lease mechanism's only purpose is to support almost-linearizable reads (assuming clocks are enough in sync) without the message round trips. 

Our raft lib doesn't actually have a concept of fully linearizable reads in the same way - we assume that reads are made directly from the FSM implemented on top of this on the leader (not fully linearizable since the leader might have lost an election during partition but not noticed yet because still inside election timeout). Where fully-consistent reads are needed (i.e. `?consistent` in Consul) we still don't process the read through raft log as some raft implementations do, instead we first get a snapshot of current leader FSM state, then run verifyLeader which essentially re-runs leader election to ensure a quorum still support the leader, and then answer the request from the snapshot that we had which is linearizable because you know the server was still the only leader at the point that snapshot was valid.

The lease mechanism in our raft lib, is setup in a similar way to the one in the thesis, but in practice it's a timer that just tells the leader how often to check heartbeats are still coming in.

So if you stop the lease here, then as soon as this method exits below (right after sending TimeoutNow, but _probably before_ the actual new leader takes over, the leaderloop above will activate the `<-lease` case, re-check the heartbeats for a quorum are still coming in, **and then rest the lease timeout again**, in other words this is kind of a no-op - it just causes the heartbeat check to happen a few ms sooner than it might.

Worst case though, if heartbeats have stopped for a while (for example because it took us a few seconds to replicate log entries above) then this will mean that immediately after we send `TimeoutNow` and return here, the leader will hit the `<-lease` case and then find it's lost lease and step down - possibly before the new leader had a chance to win it's election and take over.

Ha ha the thesis in section six says:

>  Problems due to stale reads have already been discovered in two third-party Raft implementations [45]

One of those was this one before we implemented verifyLeader 😄 .

Actually our verifyLeader mechanism technically isn't the same as the one proposed in section 6 either... We don't record the leader's commit index before and wait for it to be applied to the FSM etc. in practice given the bounds on how fast other things can happen I think it's sufficient.

The tl;dr here is that I think we can just ignore this and be no less correct than now - reads that we consider "linearizable" use a different mechanism that perform their own "heartbeats" (in thesis language, actually we call them votes in verifyLeader) anyway so we effectively don't use the `lease` as a way to assume reads are consistent.

Reply:
thank you for that explanation! removed it from the pr.

Reply:
Might be nice to keep a comment in the code to link to this for when we come back later and wonder why we didn't follow the thesis on that?